### PR TITLE
Adding struct support for semi_join

### LINKED ISSUE #7912: [FEA] Support semi_join and anti_join on structs
**Is your feature request related to a problem? Please describe.**
Currently, we have the support for equi-join on struct with `INNER_JOIN`, `LEFT_JOIN` and `FULL_JOIN`. It will be nice if we also support equi-join on struct with `LEFT_SEMI_JOIN` and `LEFT_ANTI_JOIN.`

### PR DESCRIPTION
**Pull request by hyperbolic2346**
Adding support for joining on structs for left_anti_join, left_semi_join, and left_semi_anti_join.

closes #7912

### CODE REVIEW

File: cpp/tests/join/semi_join_tests.cpp [hunk 1]
```diff
@@ -31,6 +32,80 @@
 
 template <typename T>
 using column_wrapper = cudf::test::fixed_width_column_wrapper<T>;
+using strcol_wrapper = cudf::test::strings_column_wrapper;
+using CVector        = std::vector<std::unique_ptr<cudf::column>>;
```
Comment (davidwendt) [inline cpp/tests/join/semi_join_tests.cpp:36]:
```suggestion
using column_vector        = std::vector<std::unique_ptr<cudf::column>>;
```

File: cpp/tests/join/semi_join_tests.cpp [hunk 2]
```diff
@@ -31,6 +32,80 @@
...
 };
+
+TEST_F(JoinTest, SemiJoinWithStructsAndNulls)
+{
+  column_wrapper<int32_t> col0_0{{3, 1, 2, 0, 2}};
+  strcol_wrapper col0_1({"s1", "s1", "s0", "s4", "s0"}, {1, 1, 0, 1, 1});
+  column_wrapper<int32_t> col0_2{{0, 1, 2, 4, 1}};
+  std::initializer_list<std::string> col0_names = {
+    "Samuel Vimes", "Carrot Ironfoundersson", "Detritus", "Samuel Vimes", "Angua von Überwald"};
+  auto col0_names_col = strcol_wrapper{col0_names.begin(), col0_names.end()};
```
Comment (davidwendt) [inline cpp/tests/join/semi_join_tests.cpp:49]:
```suggestion
  auto col0_names_col = strcol_wrapper{ "Samuel Vimes", "Carrot Ironfoundersson", "Detritus", "Samuel Vimes", "Angua von Überwald"};
```

File: cpp/tests/join/semi_join_tests.cpp [hunk 3]
```diff
@@ -31,6 +32,80 @@
...
+
+  column_wrapper<int32_t> col1_0{{2, 2, 0, 4, 3}};
+  strcol_wrapper col1_1({"s1", "s0", "s1", "s2", "s1"});
+  column_wrapper<int32_t> col1_2{{1, 0, 1, 2, 1}, {1, 0, 1, 1, 1}};
+  std::initializer_list<std::string> col1_names = {"Carrot Ironfoundersson",
+                                                   "Angua von Überwald",
+                                                   "Detritus",
+                                                   "Carrot Ironfoundersson",
+                                                   "Samuel Vimes"};
+  auto col1_names_col = strcol_wrapper{col1_names.begin(), col1_names.end()};
```
Comment (davidwendt) [inline cpp/tests/join/semi_join_tests.cpp:65]:
```suggestion
  auto col1_names_col = strcol_wrapper{"Carrot Ironfoundersson",
                                                   "Angua von Überwald",
                                                   "Detritus",
                                                   "Carrot Ironfoundersson",
                                                   "Samuel Vimes"};
```

File: cpp/tests/join/semi_join_tests.cpp [hunk 4]
```diff
@@ -31,6 +32,80 @@
...
+
+TEST_F(JoinTest, SemiJoinWithStructsAndNulls)
+{
+  column_wrapper<int32_t> col0_0{{3, 1, 2, 0, 2}};
+  strcol_wrapper col0_1({"s1", "s1", "s0", "s4", "s0"}, {1, 1, 0, 1, 1});
+  column_wrapper<int32_t> col0_2{{0, 1, 2, 4, 1}};
+  std::initializer_list<std::string> col0_names = {
+    "Samuel Vimes", "Carrot Ironfoundersson", "Detritus", "Samuel Vimes", "Angua von Überwald"};
+  auto col0_names_col = strcol_wrapper{col0_names.begin(), col0_names.end()};
+  auto col0_ages_col  = column_wrapper<int32_t>{{48, 27, 351, 31, 25}};
```
Comment (davidwendt) [inline cpp/tests/join/semi_join_tests.cpp:50]: Just a question here. Should these ages correlate with the names? I see Samuel Vimes is repeated in the names column.

Comment (hyperbolic2346) [reply cpp/tests/join/semi_join_tests.cpp:50]: I'm sure that originally it all did, but I've been using this as just data and not actually tying the names and ages together.

**Review by davidwendt — APPROVED** [review]
Pretty cool that enabling this just involved flattening the structs.
Only 2 suggestions. The `#include <rmm/device_vector.hpp>` is not needed and you should update the copyright year for `semi_join.cu`

**Review by ttnghia — APPROVED** [review]

**Comment by hyperbolic2346** [comment]
@gpucibot merge

**Pull request merged**